### DISCUSSION TITLE
Can CNN or RNN Deep learning models be implemented using spark-rapids-ml in Spark Rapids?

### DISCUSSION THREAD

**Discussion by leehaoun** [NVIDIA/spark-rapids#9788] [Q&A] [answered]
Hi, I'm looking to implement GPU-based distributed AI processing in Spark by applying the Rapids plugin.

I've been creating code to build deep learning models like CNN, LSTM, and RNN 

for classification directly in Tensorflow or PyTorch, which I've been using. 

I want to perform tasks involving passing 3-color images through a CNN model for classification by using RAPIDS.

Can I perform the same tasks using spark-rapids-ml? 

If not, would it be acceptable to directly import Tensorflow and Keras into Spark code for use?

**Comment by leewyang** [comment, marked as answer]
@leehaoun spark-rapids-ml is more targeted towards GPU-acceleration of the traditional ML algorithms found in [Spark MLlib](https://spark.apache.org/docs/latest/ml-guide.html).

For distributed inference using models trained in deep-learning frameworks like TensorFlow and PyTorch, you will pretty much need to directly import the model inferencing code into Spark.  We have examples of how to do this in [these examples](https://github.com/NVIDIA/spark-rapids-examples/tree/branch-23.06/examples/ML%2BDL-Examples/Spark-DL/dl_inference).

If you are trying to do distributed training on Spark, you will need to use the distributed training APIs of the respective DL frameworks combined with Spark integrations like [Horovod-on-Spark](https://horovod.readthedocs.io/en/stable/spark_include.html), [TensorFlow Distributor](https://docs.databricks.com/en/machine-learning/train-model/distributed-training/spark-tf-distributor.html), or [Pytorch Distributor](https://docs.databricks.com/en/machine-learning/train-model/distributed-training/spark-pytorch-distributor.html).

**Reply by leehaoun** [reply]
thank you for exact example.
